tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -288,15 +323,202 @@ static void write_other_stack(void)
 }
 
+static void high_prio_uthread(void)
+{
+	/* Get the priority of current thread and get higher
+	 * priority value
+	 */
+	u8_t new_prio;
+
+	new_prio = k_thread_priority_get(k_current_get()) - 2;
+
+	/* Create a user thread with priority greater than current
+	 * thread
+	 */
+	k_thread_create(&uthread_thread, uthread_stack, STACKSIZE,
+			(k_thread_entry_t)thread_body, NULL, NULL, NULL,
+			K_PRIO_PREEMPT(new_prio), K_USER | K_INHERIT_PERMS,
+			K_NO_WAIT);
+
+	/* This part of the code should not be reached */
+	zassert_unreachable("Create higher priority user thread greater"
+			" than current thread did not fault\n");
+}
+
+struct k_mem_domain app0_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app0_buf0[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT app0_buf1[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(app0_part0, app0_buf0, sizeof(app0_buf0),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+K_MEM_PARTITION_DEFINE(app0_part1, app0_buf1, sizeof(app0_buf1),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *app0_parts[] = {
+	&app0_part0,
+	&app0_part1
+};
+
+struct k_mem_domain app1_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app1_buf0[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT app1_buf1[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(app1_part0, app1_buf0, sizeof(app1_buf0),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+K_MEM_PARTITION_DEFINE(app1_part1, app1_buf1, sizeof(app1_buf1),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *app1_parts[] = {
+	&app1_part0,
+	&app1_part1
+};
+
+__kernel struct k_thread utid, ktid;
+K_THREAD_STACK_DEFINE(ustack, MEM_DOMAIN_STACK_SIZE);
+K_THREAD_STACK_DEFINE(kstack, MEM_DOMAIN_STACK_SIZE);
+
+/***************************************************************************/
+static void uthread_entry(void *destroy, void *p1, void *p2)
+{
+	if (_is_user_context()) {
+		/* check access to the memory domain app0_domain */
+		app0_buf0[0] = 1;
+
+		/* Add current thread to another memory domain app1_domain
+		 * This should generate a fault
+		 */
+		k_mem_domain_add_thread(&app1_domain, k_current_get());
+
+		/* These lines of code should not be reached */
+		zassert_unreachable("Adding a thread to multiple memory domains"
+				" did not fault\n");
+	} else{
+		zassert_unreachable("K_thread_create did not create"
+				" user thread\n");
+	}
+}
+
+/* Verify that a thread can belong to a single memory domain
+ * at any point of time
+ * Try to add a thread to two memory domains at a time
+ */
+static void thread_in_multiple_domains(void)
+{
+	k_mem_domain_init(&app0_domain, ARRAY_SIZE(app0_parts), app0_parts);
+	k_mem_domain_init(&app1_domain, ARRAY_SIZE(app1_parts), app1_parts);
+
+	/* Add current thread to memory domain app0_domain*/
+	k_mem_domain_add_thread(&app0_domain, k_current_get());
+
+	/* Create a user mode thread */
+	k_thread_create(&utid, ustack, MEM_DOMAIN_STACK_SIZE,
+			(k_thread_entry_t)uthread_entry, NULL, NULL, NULL,
+			10, K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_sleep(10);
```

Comment:
Actually, you need to k_thread_abort(k_current_get()).  Otherwise this function will return back through its call chain to test_cb(), set test_result to 0, and call k_sem_give(&test_end_signal), waking run_test() and indicating test success irrespective of the actual test result.  This is a problem whenever creating a separate thread from one of these test functions and performing the ztest_test_fail/pass from the other thread; you need to abort the current one to prevent incorrect reporting.